unit-tests/live/frames/test-fps-permutations.py

```diff
@@ -0,0 +1,304 @@
+            (mode == ALL_PAIRS and permutation.count(1) == 2))
+
+
+def get_dict_for_permutation(sensor_profiles_arr, permutation):
+    """
+    Given an array of tuples (sensor, profile) and a permutation of the same length,
+    return a sensor-profiles dictionary containing the relevant profiles
+     - profile will be added only if there's 1 in the corresponding element in the permutation
+    """
+    partial_dict = {}
+    for i, j in enumerate(permutation):
+        if j == 1:
+            sensor = sensor_profiles_arr[i][0]
+            partial_dict[sensor] = partial_dict.get(sensor, []) + [sensor_profiles_arr[i][1]]
+    return partial_dict
+
+
+def get_time_est_string(num_profiles, modes):
+    s = "Estimated time for test:"
+    details_str = ""
+    total_time = 0
+    global seconds_to_count_frames
+    global seconds_till_steady_state
+    time_per_test = seconds_to_count_frames + seconds_till_steady_state
+    for mode in modes:
+        test_time = 0
+        if mode == ALL_PERMUTATIONS:
+            test_time = math.factorial(num_profiles) * time_per_test
+            details_str += f"{math.factorial(num_profiles)} tests for all permutations"
+        elif mode == ALL_PAIRS:
+            test_time = math.comb(num_profiles, 2) * time_per_test
+            details_str += f"{math.comb(num_profiles,2)} tests for all pairs"
+        elif mode == ALL_SENSORS:
+            test_time = time_per_test
+            details_str += f"1 test for all sensors on"
+
+        details_str += " + "
+        total_time += test_time
+
+    details_str = details_str[:-3]
+    details_str = f"({details_str})"
+    details_str += f" * {time_per_test} secs per test"
+    s = f"{s} {total_time} secs ({details_str})"
+    return s
+
+
+def get_tested_profiles_string(sensor_profiles_dict):
+    s = ""
+    for sensor in sensor_profiles_dict:
+        for profile in sensor_profiles_dict[sensor]:
+            s += sensor.name + " / " + profile.stream_name() + " + "
+    s = s[:-3]
+    return s
+
+
+#############################################
+# ---------- Core Test Functions ---------- #
+#############################################
+def check_fps(expected_fps, fps_measured):
+    all_fps_ok = True
+    for key in expected_fps:
+        res = check_fps_pair(expected_fps[key], fps_measured[key])
+        if not res:
+            all_fps_ok = False
+            log.e(f"Expected {expected_fps[key]} fps, received {fps_measured[key]} fps in sensor {key.name}")
+    return all_fps_ok
+
+
+def generate_functions(sensor_profiles_dict):
+    """
+    Creates callable functions for each sensor to be triggered when a new frame arrives
+    Used to count frames received for measuring fps
+    """
+    sensor_function_dict = {}
+    for sensor_key in sensor_profiles_dict:
+        def on_frame_received(frame, key=sensor_key):
+            global count_frames
+            if count_frames:
+                sensor_profiles_dict[key] += 1
```

Comment:
Are you sure we can use 2 parames callback for the frame? interesting.
We call it with 1 parameter in the SDK.
Are you sure no lock is needed and that python does not need to use locks?

Reply:
It seems like we can, maybe because one of them has a default value.
And i think locks are needed, I'll add them.